README edits (#43)
astro-projects/astro: README.md

</h3>
<br/>

> # Overview

Author: I think that we need a more straightforward introduction to what `astro` is. I took my best shot at creating an Overview section here, but please add more / less detail depending on how you view this project. I also thought including a before/ after section would make it even more clear what value `astro` adds to your Airflow project.

> The `astro` library is a collection of SQL and dataframe modules that you can use to write tasks without needing to define Airflow connections and metadata for each task. By design, `astro` modules automatically pass database contexts to your tasks, meaning that you can focus on writing Python and leave metadata definitions for load time.

Author: The `astro` library is a suite of tools for writing ETL and ELT workflows in Airflow. It lets SQL engineers focus on writing SQL, Python engineers focus on writing Python, and lets all engineers focus on data engineering instead of configuration. By design, `astro` modules automatically pass database contexts to your tasks, meaning that you can focus on writing code and leave metadata definitions for load time.

- **Without `astro`**: Database connections and context are defined in operators. This metadata does not pass automatically from task to task, meaning that you have to redefine it for each new task/ operator. SQL queries must be static. 

> - **With `astro`**: Databse connections and context are defined at load time. Metadata is automatically passed into functions, meaning that you can write generic queries in your task an run these queries in various contexts without knowing the context ahead of time. SQL queries can be templated with JINJA.

Author: ```suggestion
- **With `astro`**: Database connections and context are defined at load time. Metadata is automatically passed into functions, meaning that you can write generic queries in your task an run these queries in various contexts without knowing the context ahead of time. SQL queries can be templated with JINJA.
```

    ```shell script
    pip install astro-projects
    ```

    Alternatively, you can add `astro-projects` to your `requirements.txt` file.

2. Set the following environment variable so that `astro` can pass table objects between tasks:

    ```shell script
    AIRFLOW__CORE__ENABLE_XCOM_PICKLING=True
    ```

#  Using Astro as a SQL Engineer

## Schemas

By default, running `astro` for a given database creates a schema called `tmp_astro` in the database. This default behavior assumes that you have permissions to create schemas on the fly, and that there is only one user adding/removing from this schema.

For production usage, we recommend that you work with your database administrator to create a shared schema for temporary tables. This schema
can be shared across multiple users, but should be created with security in mind (e.g. don't place high security data in a shared schema).

Once you create this schema, an Airflow admin can replace `tmp_astro` by setting `AIRFLOW__ASTRO__SQL_SCHEMA="<new-temp-schema>"`, or by setting the following in `airflow.cfg`:

```bash
[astro]
sql_schema=<new-temp-schema>
```

## Setting up SQL files

> When writing out a SQL DAG using `astro`, you can think of each SQL file as its own task. For example, if you wanted

Author: ```suggestion
When writing out a SQL DAG using `astro`, you can think of each SQL file as its own Airflow task. For example, if you wanted
```

> to aggregate orders, aggregate customers, and then join customers and orders you could havesh

Author: ```suggestion
to aggregate orders, aggregate customers, and then join customers and orders, you could have the following directory of files:
```

```

> |

Author: Expose top level directory here to match up with the examples later on?

Author: ingest-models

-- customers_table.sql
-- orders_table.sql
-- join_customers_and_orders.sql
```

> In each of these SQL files, you can create a table by having a standard `SELECT` statement. `astro` handles creating all of the temporary tables.

Author: Not clear what table you're creating here. Is it the table that becomes our `output_table` in later suggestions?

Author: Make it clear that the table itself is automatically generated and can be referenced in other SQL queries by the data dependency alone

```sql
# join_customers_and_orders.sql
SELECT c.customer_id, c.source, c.region, c.member_since,
        CASE WHEN purchase_count IS NULL THEN 0 ELSE 1 END AS recent_purchase
        FROM orders c LEFT OUTER JOIN customers p ON c.customer_id = p.customer_id
```

### Defining metadata

> Once your SQL is working as expected, you might want to define the database and schema for the query when the query itself runs.

Author: ```suggestion
Once your SQL is working as expected, you might want to define the query's database and schema during its runtime.
```

> To expose this functionality while keeping your SQL easy to run in your favorite SQL notebook, 

Author: ```suggestion
To configure this for Airflow while keeping your SQL easy to run in your favorite SQL notebook, 
```

### Defining dependencies

When running SQL queries in Airflow DAGs, you need to define dependencies that break up your SQL into

> multiple, reproduceable steps. We offer two ways to define dependencies within an `astro` SQL file:

Author: ```suggestion
multiple, reproducible steps. We offer two ways to define dependencies within an `astro` SQL file:
```

> The first way to define a dependency is by defining a variable via the `template_vars` frontmatter argument. We set `template_vars` so that our `join_customers_and_orders` query cannot run until both our `customers_table` and `agg_orders` queries finish. This data dependecy is equivalent task dependency in our DAG. The only difference is that we're defining it directly in our SQL instead of using Airflow's dependency operators. 

Author: ```suggestion
The first way to define a dependency is by defining a variable via the `template_vars` frontmatter argument. We set `template_vars` so that our `join_customers_and_orders` query cannot run until both our `customers_table` and `agg_orders` queries finish. This data dependency is equivalent to a task dependency in our DAG. The only difference is that we're defining it directly in our SQL instead of using Airflow's dependency operators. 
```

> The first way to define a dependency is by defining a variable via the `template_vars` frontmatter argument. We set `template_vars` so that our `join_customers_and_orders` query cannot run until both our `customers_table` and `agg_orders` queries finish. This data dependency is equivalent to a task dependency in our DAG. The only difference is that we're defining it directly in our SQL instead of using Airflow's dependency operators. 

Author: We never state a second way?

> The first way to define a dependency is by defining a variable via the `template_vars` frontmatter argument. We set `template_vars` so that our `join_customers_and_orders` query cannot run until both our `customers_table` and `agg_orders` queries finish. This data dependency is equivalent to a task dependency in our DAG. The only difference is that we're defining it directly in our SQL instead of using Airflow's dependency operators. 

Author: Leave as "the one way"

        FROM orders c LEFT OUTER JOIN customers p ON c.customer_id = p.customer_id
```

> Here is a list of currently supported frontmatter argument:

Author: ```suggestion
Here is a list of supported frontmatter arguments:
```
Might make sense to have this in a "Reference" section outside of the guide component of this doc.

    ingest_models = aql.render(dir_path + "/ingest_models", orders_table=raw_orders)
```

Once the `render` function completes, it returns a dictionary of all of the models based on the sql file name.
If you have a sql file named `join_orders_and_customers.sql`, then the result would be stored in `model["join_orders_and_customers]`.

### Passing on tables to subsequent tasks

> The following DAG is an example of passing a single model to a subsequent rendering (perhaps you want to separate your ingest and transforms):

Author: The `**` example seems to be an exact copy of the previous example, pretty sure that's a typo.

    )
```

You can even pass the resulting tables into a python function that uses the `astro.dataframe` to

> automatically convert your table into a dataframe (but we'll go into more detail on that further down).

Author: ```suggestion
automatically convert your table into a dataframe. We'll discuss how to do this more in "Using Astro as a Python Engineer".
```

>     )

Author: **ingest_models,

> For those who don't want to store their transformations in external SQL files or who want to create tranformation

Author: ```suggestion
For those who don't want to store their transformations in external SQL files or who want to create transformation
```

functions that are extendable and importable, we offer a rich python API that simplifies the SQL experience for the python engineer!

## Setting Input and Output Tables

Before we can complete any transformations, we need to define a way to get our tables in and out of Airflow. We can do this by defining either `Table` or `TempTable` objects in the `input_table` and `output_table` parameters of our table instantiations.

### The Table class

> To instantiate a table or bring in a table from a database into the `astro` ecosystem, you can pass a `Table` object into the class. This Table object will contain all of the metadata that's necessary for handling table creation between tasks. Once you define at in the beginning of your pipeline, `astro` can automatically pass that metadata along to downstream tasks.

Author: ```suggestion
To instantiate a table or bring in a table from a database into the `astro` ecosystem, you can pass a `Table` object into the class. This Table object will contain all of the metadata that's necessary for handling table creation between tasks. After you define a Table's metadata in the beginning of your pipeline, `astro` can automatically pass that metadata along to downstream tasks.
```

> In the following example, we define the our table in the DAG instantiation. In each subsequent task, we only pass in an input table argument because `astro` automatically passes in the additional context from our original `input_table` parameter.

Author: ```suggestion
In the following example, we define our table in the DAG instantiation. In each subsequent task, we only pass in an input table argument because `astro` automatically passes in the additional context from our original `input_table` parameter.
```

    )
```

### Raw SQL

Most ETL use cases can be addressed by cross-sharing task outputs, as shown above with `@aql.transform`. If you need to perform SQL operation that doesn't return a table but might take a table as an argument, you can use `@aql.run_raw_sql`. 

```python
@aql.run_raw_sql
def drop_table(table_to_drop):
    return "DROP TABLE IF EXISTS {table_to_drop}"
```

## Putting it All Together 

> The following is a full example DAG of a SQL + Python workflow using `astro`.

Author: Can we add some more explanation to what this DAG is doing? What features from before are we using here?

Author: Pulling data from S3, running SQL transformations to merge with existing data, and moving that result into a dataframe so that you can do complex work on it with Python / ML
